ManagedMediaSource content should be evicted whenever it is under memory pressure

jyavenard opened the pull request:
#### fee371950f6239d24a6734e461dfcedd4a1daa8a
<pre>
ManagedMediaSource content should be evicted whenever it is under memory pressure
<a href="https://bugs.webkit.org/show_bug.cgi?id=252938">https://bugs.webkit.org/show_bug.cgi?id=252938</a>
rdar://105908445

Reviewed by Youenn Fablet.

Let the MediaElement control when eviction under memory pressure should
occur.

* LayoutTests/media/media-source/media-managedmse-memorypressure-expected.txt:
* LayoutTests/media/media-source/media-managedmse-memorypressure.html:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::purgeBufferedDataIfPossible):
* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseCriticalMemory):

Canonical link: <a href="https://commits.webkit.org/260871@main">https://commits.webkit.org/260871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8716e641bfd50a7f38aa904f7fc31fa93432d990

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109761 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18876 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42498 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1231 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118880 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113712 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20347 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10073 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102029 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115509 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15147 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98378 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43369 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30031 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85162 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11602 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31373 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12254 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8312 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50982 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7537 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14004 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -8643,6 +8643,17 @@ void HTMLMediaElement::purgeBufferedDataIfPossible()
+        return paused();
+    }();
+
+    if (isPausedAndNoMSE)
```
s/isPausedAndNoMSE/isPausedAndNoMSE()/

jyavenard replied:
done.

youennf replied:
Why was the test passing? Should we add a test that checks the false case?

jyavenard replied:
why would it not pass?. The lambda was evaluated before and the result set to the value.

jyavenard replied:
BTW: why you would want to evaluate the lambda in the if statement, rather than when creating the `isPausedAndNoMSE` variable is puzzling to me.

jyavenard replied:
actually, on second thought, I like the original way better, and avoid to have to keep the lambda for longer than necessary

jyavenard replied:
> Should we add a test that checks the false case?

in the non MSE case, how the data is evicted (and if it's going to be) is managed by AVFoundation I'm not sure how this could be tested in a guaranteed outcome

youennf replied:
Oh I misread the code, sorry about the noise. Yes this is correct as is.

jyavenard committed:
ManagedMediaSource content should be evicted whenever it is under memory pressure
https://bugs.webkit.org/show_bug.cgi?id=252938
rdar://105908445

Reviewed by Youenn Fablet.

Let the MediaElement control when eviction under memory pressure should
occur.

* LayoutTests/media/media-source/media-managedmse-memorypressure-expected.txt:
* LayoutTests/media/media-source/media-managedmse-memorypressure.html:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::purgeBufferedDataIfPossible):
* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseCriticalMemory):

Canonical link: https://commits.webkit.org/260871@main
